tedana/viz.py

```diff
@@ -0,0 +1,257 @@
+        Array with columns denoting (1) index of component, (2) Kappa score of
+        component, (3) Rho score of component, (4) variance explained by
+        component, and (5) normalized variance explained by component
+    mmix : (C x T) array_like
+        Mixing matrix for converting input data to component space, where `C`
+        is components and `T` is the same as in `data`
+    n_vols : :obj:`int`
+        Number of volumes in original time series
+    acc : :obj:`list`
+        Indices of accepted (BOLD) components in `mmix`
+    rej : :obj:`list`
+        Indices of rejected (non-BOLD) components in `mmix`
+    midk : :obj:`list`
+        Indices of mid-K (questionable) components in `mmix`
+    empty : :obj:`list`
+        Indices of ignored components in `mmix`
+    ref_img : :obj:`str` or img_like
+        Reference image to dictate how outputs are saved to disk
+
+    """
+
+    # regenerate the beta images
+    ts_B = model.get_coeffs(ts, mmix, mask)
+    ts_B = ts_B.reshape(ref_img.shape[:3] + ts_B.shape[1:])
+    # Mask out zeros
+    ts_B = np.ma.masked_where(ts_B == 0, ts_B)
+
+    # Get repitition time from ref_img
+    tr = ref_img.header.get_zooms()[-1]
+
+    # Start making plots
+    if not os.path.exists('figures'):
+        os.mkdir('figures')
+
+    # Create indices for 6 cuts, based on dimensions
+    cuts = [ts_B.shape[dim] // 6 for dim in range(3)]
+
+    for compnum in range(0, mmix.shape[1], 1):
+
+        allplot = plt.figure(figsize=(10, 9))
+        ax_ts = plt.subplot2grid((5, 6), (0, 0), rowspan=1, colspan=6,
+                                 fig=allplot)
+        if compnum in acc:
+            line_color = 'g'
+        elif compnum in rej:
+            line_color = 'r'
+        elif compnum in midk:
+            line_color = 'm'
+        else:
+            line_color = 'k'
+
+        ax_ts.plot(mmix[:, compnum], color=line_color)
+
+        # Title will include variance from comptable
+        comp_var = "{0:.2f}".format(comptable.iloc[compnum][3])
+        plt_title = 'Comp. {}: {}% variance'.format(compnum, comp_var)
+        title = ax_ts.set_title(plt_title)
+        title.set_y(1.5)
+        ax_ts.set_xlabel('TRs')
+        ax_ts.set_xbound(0, n_vols)
+        # Make a second axis with units of time (s)
+        max_xticks = 10
+        xloc = plt.MaxNLocator(max_xticks)
+        ax_ts.xaxis.set_major_locator(xloc)
+
+        ax_ts2 = ax_ts.twiny()
+        ax1Xs = ax_ts.get_xticks()
+
+        ax2Xs = []
+        for X in ax1Xs:
+            # Limit to 2 decimal places
+            seconds_val = round(X * tr, 2)
+            ax2Xs.append(seconds_val)
+        ax_ts2.set_xticks(ax1Xs)
+        ax_ts2.set_xbound(ax_ts.get_xbound())
+        ax_ts2.set_xticklabels(ax2Xs)
+        ax_ts2.set_xlabel('seconds')
+
+        # Set range to ~1/10th of max beta
+        imgmax = ts_B[:, :, :, compnum].max() * .1
+        imgmin = ts_B[:, :, :, compnum].min() * .1
+
+        for imgslice in range(1, 6, 1):
```

Comment:
The default step in `range` is 1, so this is equivalent to `range(1, 6)` !

Reply:
Actually, what if we re-write this section like this (forgiving indentation differences !):

```suggestion
for cut in cuts:
    imgslice = [i for i in range(1, 6)]
    for i in imgslice:
        ax = plt.subplot2grid((5, 6), (1, imgslice - 1), rowspan=1, colspan=1)
        ax.axis('off')
        ax.imshow(np.rot90(ts_B[imgslice * cut, :, :, compnum]),
                  vmin=imgmin, vmax=imgmax, aspect='equal',
                  cmap='coolwarm')
```

That way we don't repeat the same block three times. WDYT ?

Reply:
The trouble is that the np.rot90 is only needed for two of the figures - the other one seems to be the right way up to start with - so would lead to a silly figure. Is there a more clever way of handling that issue? This is, for the moment, dodging the issue of whether is np.rot90 is always needed - that is pending testing is A-P, R-L, etc collection.

Reply:
Ah, I see ! Sorry to have missed that. We could make it into a boolean like this:

```suggestion
for idx, cut in enumerate(cuts):

    if idx in [0, 1]:  # for the first two dimensions
        rotate = True

    imgslice = [i for i in range(1, 6)]
    for i in imgslice:
        ax = plt.subplot2grid((5, 6), (1, imgslice - 1), rowspan=1, colspan=1)
        ax.axis('off')

        if rotate:
            ax.imshow(np.rot90(ts_B[imgslice * cut, :, :, compnum]),
                      vmin=imgmin, vmax=imgmax, aspect='equal',
                      cmap='coolwarm')
        else:
            ax.imshow(ts_B[imgslice * cut, :, :, compnum],
                      vmin=imgmin, vmax=imgmax, aspect='equal',
                      cmap='coolwarm')

    rotate = False  # by default, don't rotate
```

Doesn't end up saving a ton of space, but I find it a little clearer, I think !

Reply:
Another, nicer way to do this:

```python
for idx, cut in enumerate(cuts):

    for imgslice in range(1, 6):
        ax = plt.subplot2grid((5, 6), (1, imgslice - 1), rowspan=1, colspan=1)
        ax.axis('off')

        to_plot = ts_B[imgslice * cut, :, :, compnum]
        if idx in [0, 1]:  # for the first two dimensions
            to_plot = np.rot90(to_plot)  # rotate the plotted slices by 90-deg

        ax.imshow(to_plot, vmin=imgmin, vmax=imgmax,
                  aspect='equal', cmap='coolwarm')
```

Reply:
I'll try and adapt the latter example, however, there are two large issues. The slice needs to occur across different indices, and the subplots needs to be shift on the same basis.

Reply:
I've found what may be a compromise - requires 3 `if` statements though, in order to make the cuts along the right axes, and rotate if needed. Single plot statement though, so thats technically a move in the right direction.